### PR TITLE
CIP-0381? | Plutus support for pairings over curve Bls12-381

### PR DESCRIPTION
**Pull request by iquerejeta**
This CIP proposes to include plutus support for pairings over curve Bls12-381

### CODE REVIEW

**Comment by L-as** [comment]
I don't think this approach is viable. Even if you add support for pairings, executing the algorithms in question in the CEK machine would be horrifically slow. It is a much better idea to add support for a single universal proof system, and I highly recommend Halo 2. The main blocker is the API for recursive proofs not being done yet, but that should change soon AFAICT.

**Comment by iquerejeta** [comment]
I've added some estimates of the number of operations for each of the different application listed in the CIP. Even for Plonk verification I get estimates that could fit within the plutus execution budget. In particular the significant operations in a verification are 
* 6 hash computations
* 1 vanishing polynomial (1 scalar exponantiation)
* Quotient polynomial evaluation (1 scalar inverse)
* First part of batch poly commitment (7 scalar mults and 6 additions in G1)
* Fully batched poly commitment (7 scalar mults and 8 additions over G1)
* Group encoded batch verification (1 scalar mult over G1)
* Batch-validate all evaluations (3 scalar mults and 4 additions over G1, plus two pairing evaluations + equality check)

@L-as, could you elaborate on why the CIP is not a viable approach in general for the other primitives please? And also, could you elaborate on why you think that Halo 2 (with IPA) is the best choice for a universal proof system (over Plonk, for instance)? I believe that verification of relatively complex circuits with Halo2 will most probably exceed plutus execution budget. 

Thanks!

File: CIP-0037/CIP-0037.md
Comment (michaelpj) [inline CIP-0037/CIP-0037.md:158]: k?

Comment (michaelpj) [reply CIP-0037/CIP-0037.md:158]: and below

**Comment by iquerejeta** [comment]
Ok, yes. I will add that as well. It's true that the numbers with respect to ed25519 might not be very helpful.

File: CIP-0037/CIP-0037.md
Comment (michaelpj) [inline CIP-0037/CIP-0037.md:34]: Github markdown doesn't support latex maths, so these look pretty bad throughout. Maybe we're okay just asking people to render the latex in their heads, but it might be nice to try and use ascii.

Comment (L-as) [reply CIP-0037/CIP-0037.md]: You can use inline SVG IIRC.

**Comment by L-as** [comment]
> And also, could you elaborate on why you think that Halo 2 (with IPA) is the best choice for a universal proof system (over Plonk, for instance)? I believe that verification of relatively complex circuits with Halo2 will most probably exceed plutus execution budget.

If we are to support Halo 2 directly in the interpreter, it would have a budget defined by the interpreter. It wouldn't be extremely fast, but we are free to choose whatever cost model we want when putting it into the interpreter. In practice, it would probably be half the CPU and memory budget + 1 or something similar.

As for why Halo 2, there isn't really any competitor. The four important properties are recursive proofs, trustlessness, open source license, and quality code base. There isn't any alternative to Halo 2 that accomplishes all of this.

**Comment by iquerejeta** [comment]
> As for why Halo 2, there isn't really any competitor.

Plonk is being used in several applications, so I would consider it as a competitor. 

> The four important properties are recursive proofs, trustlessness, open source license, and quality code base.

Agree, however, plonk trusted setup is universal (meaning that it has to be performed only once) and is updatable which, IMO, it is enough for a lot of scenarios (maybe not all). What I'm concerned of Halo2 is the time it takes to verify a proof. It is my understanding that the plutus execution budget is 5-10ms. I don't have recent benchmarks at hand, but at least what was published in the Halo paper indicates that verifying a circuit of 2^11 constraints (pretty quite limiting) takes 0,1 seconds, and growth is linear. And verifying 2^17 constraints takes close to 1 second. I know these numbers are definitely better using Halo2, but I couldn't find any benchmarks. Maybe you have some numbers you can share to support the statement that it will be around half the CPU budget? For reference, using [plonk](https://github.com/ZK-Garage/plonk), times are 4,13ms and 5,15ms respectively.

**Comment by michaelpj** [comment]
> It wouldn't be extremely fast, but we are free to choose whatever cost model we want when putting it into the interpreter. In practice, it would probably be half the CPU and memory budget + 1 or something similar.

We certainly can't just pick whatever cost model we like. It's important that it's at least somewhat accurate, because ultimately those limits are what stop us from blowing out our block diffusion times. If Halo 2 takes 2ms to verify something, it's going to have to be costed like that.

File: CIP-0037/CIP-0037.md
Comment (markulf) [inline CIP-0037/CIP-0037.md:1]: Suggestions: 

Also expose non-pairing curves using same API (except for pairing).

**Comment by L-as** [comment]
@iquerejeta I think I agree now that we should just add support for the curve to UPLC now instead of adding support for the verification directly. However, do the Pasta curves also work for your use case?

**Comment by iquerejeta** [comment]
The support that we will be including in this CIP is only for BLS12-381 curve, not the pasta curves. However, other curves could be considered if we see that they provide attractive properties. I'm guessing that the main attractive point of pasta curves is its recursive nature, right? If that is the case, considering plonky2 might be worth it, that combines ZKPs using FRI commitments (for speed in producing the proof), and batches them using a Plonk proof with KZG10 commitments (for succinctness in the proof size). I'm no expert in the field, but I believe this would provide attractive performance when batching several proofs, and would be compatible with the introduction of BLS12-381 curves. I would be more than happy in having a more in-depth discussion on this and the general topic of zkps and roll ups.

**Comment by L-as** [comment]
FYI even the smallest Plonky2 proof is too big to fit in a single transaction AFAICT. In any case, we should definitely talk about this. Feel free to set up something.

File: CIP-0037/CIP-0037.md
Comment (davidnevadoc) [inline CIP-0037/CIP-0037.md:319]: The specific wire connections do not affect the verifier algorithm complexity. The connections are enforced as copy constraits with a permutation argument whose complexity is the same no matter the permutation we are using.

File: CIP-0037/CIP-0037.md
Comment (davidnevadoc) [inline CIP-0037/CIP-0037.md:321]: Better use a reference to the [paper](https://eprint.iacr.org/2019/953.pdf) as it is more updated. 
In particular the current version has a new way of computing the linearisation polynomial that saves some operations for the verifier.

File: CIP-0037/CIP-0037.md
Comment (davidnevadoc) [inline CIP-0037/CIP-0037.md:328]: No need to compute the quotient evaluation in the newest version.

File: CIP-0037/CIP-0037.md
Comment (davidnevadoc) [inline CIP-0037/CIP-0037.md:329]: Now: 10 scalar mults and 9 additions in G_1

File: CIP-0037/CIP-0037.md
Comment (davidnevadoc) [inline CIP-0037/CIP-0037.md:330]: I've counted 9 scalar mults in the previous in the previous version.
Now: 6 scalar mults and 5 additions in G_1

File: CIP-0037/CIP-0037.md
Comment (davidnevadoc) [inline CIP-0037/CIP-0037.md:335]: This is a good estimate as a reference, but we have to expect higher costs because this is the most basic version of Plonk and real life implementations will likely be more complex. 
For example adding lookups gates introduces tables (with its corresponding commitments), an extra permutation argument (increases the number of scalar multiplications and additions) and the need for new challenges (more hashes).

**Review by davidnevadoc — CHANGES_REQUESTED** [review]
I have added some comments in the Plonk section with suggestions to make it a bit more precise.
In general, I think this is a very good and useful proposal! :)

File: CIP-0037/CIP-0037.md
Comment (iquerejeta) [reply CIP-0037/CIP-0037.md:330]: I count 9 and 9 in the first part of the poly com, and 5 and 5 in the second part. Thanks for the suggestion of referencing the paper, will do that 👍

File: CIP-0037/CIP-0037.md
Comment (CPerezz) [reply CIP-0037/CIP-0037.md:335]: Blinder costs should also be considered as well as the custom gates that any implementation uses.

From my perspective, plonk is "completely agnostic to this PR". In the sense that any project will be free to implement it's own Plonk Verifier contract/lib with their own custom gate design picks for their use-cases.
What I think would be really valuable is to first make sure that this is indeed doable for `vanilla-plonk` or `turbo-plonk` (excluding custom gates) as we still don't have any idea of how much this execution will cost or even if it is possible to execute all that stuff inside a contract.

Comment (iquerejeta) [reply CIP-0037/CIP-0037.md:335]: Thanks. Indeed, as you say this CIP is agnostic to what the pairings are used for. This list served as an estimate of some of the use cases that could leverage pairings. On your second point, I complete agree as well. Once we get the pairings running on a test version of plutus (and we have the functions costed), we'll see how it performs for verifying a vanilla plonk, and how much of the execution budget we spend.

**Review by michaelpj — COMMENTED** [review]
I think this is starting to look viable and we should maybe move to a prototype implementation.

File: CIP-0037/CIP-0037.md
Comment (iquerejeta) [reply CIP-0037/CIP-0037.md:152]: What we had before (of only the equality check) allowed us to compare two pairing evaluations, e.g. `e(A,B) =? e(C,D)` where `e(,)` is the pairing function. However, in some cases, we want to have more complex linear relations of pairing evaluations, e.g. `e(A, B) * e(C, D) =? e(E, F) * e(G, H)`. The benefit of using this type of representation comes here. If we did not use that, we would take around 700-800 us for each pairing, resulting in ~3ms for the whole. Instead, by using this representation we expect around 1,6ms. 

I'll clarify this in the section.

File: CIP-0037/CIP-0037.md
Comment (michaelpj) [reply CIP-0037/CIP-0037.md]: Ignore me, you can't lambda-encode it, that's a higher-order function. This is actually a bit tricky.

Comment (iquerejeta) [reply CIP-0037/CIP-0037.md]: But then, if we want to handle the error, what would be your recommendation?

**Comment by L-as** [comment]
One very important thing to note here, is that we're likely going to add more curves in the future, notably the Pasta (Pallas and Vesta) Curves, secp256k1 (the curve).

The core question I have then is, should we for each curve add the same primitives again separately?
Should basic primitives like addition work for many curves (and even integers)?

File: CIP-0037/CIP-0037.md
Comment (L-as) [inline CIP-0037/CIP-0037.md:118]: Why not just `deserialiseG1 :: ByteString -> BuiltinList Bls12_381G1Element`, where the output is empty if it errs and otherwise a single element? We don't really care about error handling too much on-chain.
You could also just make it partial, `deserialiseG1 :: ByteString -> Bls12_381G1Element` looks reasonable to me.

File: CIP-0037/CIP-0037.md
Comment (L-as) [inline CIP-0037/CIP-0037.md:95]: Shouldn't we use BLAKE2 as is common in Cardano AFAICT? There are a lot of reasons to avoid SHA2, see https://github.com/zcash/zcash/issues/706.

File: CIP-0037/CIP-0037.md
Comment (L-as) [reply CIP-0037/CIP-0037.md:180]: So essentially, you'd add `BLS12_381_full_pairing`? Shouldn't `BLS12_381_pairing` just have a different name then?

File: CIP-0037/CIP-0037.md
Comment (L-as) [inline CIP-0037/CIP-0037.md:100]: Idea that @morganthomas told me, we could have each of these generic group operations be indexed with the type of the group essentially, so `add :: (g :: G) -> g -> g -> g` in a dependently typed language, where `g` is internally represented as an integer. In the implementation, the interpreter would then err for invalid choices of `g`.

Comment (L-as) [reply CIP-0037/CIP-0037.md:100]: This might be better for supporting other curves in the future, what do you think @michaelpj ?

File: CIP-0037/CIP-0037.md
Comment (L-as) [reply CIP-0037/CIP-0037.md:118]: I can't think of a single case where handling the error is necessary.

File: CIP-0037/CIP-0037.md
Comment (iquerejeta) [reply CIP-0037/CIP-0037.md:180]: Yes, right. If there is the possibility that in the future we would add the full pairing, we should name it differently here. Maybe something like `BLS12_381_ppairing_ml` (standing for partial pairing miller loop), and then have clear documentation of the function.

File: CIP-0037/CIP-0037.md
Comment (iquerejeta) [reply CIP-0037/CIP-0037.md:100]: I think having a generic strategy is worth considering, though we need the opinion of @michaelpj here. I don't think we will add support for a lot of pairing friendly curves though, as this moves somewhat slowly. But in the mid/long term, this might be the case. 

My understanding is that the motivation for this is to include non-pairing friendly curves (such as pasta curves). My concern is that this generic strategy will become quite complex if we want to support all kind of EC families, and I'm not sure I see a huge benefit over simply adding support for different curves. For instance, do we want the interpreter to be aware of the cycles of the pasta curves (or any cycles of pairing friendly curves)? how would the pairing fit in such a generic framework? Would we have something like `BLS12_381_pairing :: (g1 :: G, g2::G) -> g1 -> g2 -> something`? (sorry if my notation is incorrect, I'm not a haskell developer). How would we represent the target group? 

I think that having a generic strategy that supports pairing friendly and non-pairing friendly curves would be beneficial, as we could also use it to expose curve operations over secp256k1 or curve25519.

File: CIP-0037/CIP-0037.md
Comment (iquerejeta) [reply CIP-0037/CIP-0037.md:95]: We are using the recommendation of the standard for BLS12 381, which is also what is implemented in the blst library. BLS signatures depend on this choice, and the goal is to follow the standard to be compatible with as many implementations of BLS signatures over BLS12 381. 

While Blake is the common and default choice in  Cardano, there are some exceptions. These exceptions are made to follow the standard specifications and make Cardano verification procedures compatible with other implementations. Notable exceptions are ed25519 and VRF (where both use SHA512 to compute the challenge).

File: CIP-0037/CIP-0037.md
Comment (michaelpj) [reply CIP-0037/CIP-0037.md:118]: Why can't we just fail the whole thing if deserialization fails? Surely this indicates malformed input of some kind?

File: CIP-0037/CIP-0037.md
Comment (michaelpj) [reply CIP-0037/CIP-0037.md:100]: I don't think we can do this easily. Firstly, we don't have the dependent typing. Secondly, if we added new curves to the support here, that would mean that some builtin calls which had previously failed would start succeeding, which is a visible semantics change.

Comment (L-as) [reply CIP-0037/CIP-0037.md:100]: @michaelpj I agree in that it *is* a change to the semantics, but isn't that true for *every* change? A transaction that attempts to use Plutus V2 will fail now, but won't in the future, no?

wrt. dependent typing, that was simply to make it clear what the intent was. You could have the signature be `add :: Int -> G -> G -> G` where `G` is a sum type that represents all the possible choices.

Comment (L-as) [reply CIP-0037/CIP-0037.md:100]: @michaelpj I agree in that it *is* a change to the semantics, but isn't that true for *every* change? A transaction that attempts to use Plutus V2 will fail now, but won't in the future, no?

wrt. dependent typing, that was simply to make it clear what the intent was. You could have the signature be `add :: Int -> G -> G -> G` where `G` is a sum type that represents all the possible choices.

File: CIP-0037/CIP-0037.md
Comment (L-as) [reply CIP-0037/CIP-0037.md:118]: I am very much for just failing the whole thing.

File: CIP-0037/CIP-0037.md
Comment (michaelpj) [reply CIP-0037/CIP-0037.md:100]: It's a change to the semantics of *existing scripts*. That we can't do easily. And sadly even failures are visible, because you can post a failed script to the chain if it's a phase 2 failure.

Comment (L-as) [reply CIP-0037/CIP-0037.md:100]: @michaelpj Well the Plutus version could determine what indices are valid. That would solve the issue AFAICT.

Comment (michaelpj) [reply CIP-0037/CIP-0037.md:100]: If you were willing to do a new Plutus version every time. But we're expecting those to be very infrequent, we can normally add builtins without needing one.

Comment (L-as) [reply CIP-0037/CIP-0037.md:100]: Well that sounds fine to me honestly. We don't want that many anyway.
However, can you think of a better solution rather than this? Is it more sensible to duplicate the operations for each curve? Are we duplicating type checking functionality? What is the most efficient? Is having that many builtins (O(n) for n curves) practical? Could the implementation be simpler if they were the same builtin?

Comment (iquerejeta) [reply CIP-0037/CIP-0037.md:100]: They would certainly not have identical performance characteristics. And as I said above, different curves (or more precisely, groups) will even have different functions. For instance, if we were to include operations over the group defined over secp256k1, we would only need `add`, `mult` and `neg` functions once (together with serialization and equality checks), but there would be no concept of pairing, `G1` or `G2`, and target group. If, on the other hand, we were to include operations over the groups defined over the pasta curves, we would need those operations together with some definition of `cycles` (something not covered in this CIP). I'm still not sure that having a single generic API covering all types of groups over ECs (including non-prime order groups, pairing friendly curves and cycles over different curves) will be simpler, provided that we don't plan on supporting a very big number of curves.

Comment (L-as) [reply CIP-0037/CIP-0037.md:100]: I agree with this, albeit it's unfortunate.

**Comment by iquerejeta** [comment]
I believe we are in a position to start prototyping. Not sure who should approve the CIP, maybe @KtorZ ? I guess we need to adapt the number of the CIP, but I guess we can leave to the last moment.

File: CIP-0037/CIP-0037.md
Comment (L-as) [reply CIP-0037/CIP-0037.md:183]: @michaelpj Thoughts on this? Should we try to make a complete API or should we add it in another Plutus language version in the future if someone needs it?

**Comment by L-as** [comment]
You can start working on an implementation before acceptance/rejection FWIW.

File: CIP-0037/CIP-0037.md
Comment (L-as) [inline CIP-0037/CIP-0037.md:107]: Shouldn't this be `ByteString` instead of `ByteArray`?

File: CIP-0037/CIP-0037.md
Comment (L-as) [inline CIP-0037/CIP-0037.md:112]: typo

**Comment by L-as** [comment]
what happened to PR name?

File: CIP-0037/CIP-0037.md
Comment (L-as) [reply CIP-0037/CIP-0037.md:186]: bump

**Comment by KtorZ** [comment]
@L-as regarding PR's name: CIP numbers are assigned on a best-effort basis. It sometimes happen that two CIPs are submitted along the same time and authors allocate the same numbers. That's why encourage authors not to assign numbers themselves, we do it as part of the housekeeping. Plus, a number is only truly settled once the CIP is accepted and merged. Hence the `?` on candidates CIPs. This is a summarized on the main README.

File: CIP-0037/CIP-0037.md
Comment (michaelpj) [reply CIP-0037/CIP-0037.md:186]: It's hard to say in advance. I think the efficiency arguments are compelling, but it would indeed be good to identify clearly what the functions that we are adding do, e.g. naming and documenting them appropriately. That will also reduce confusion if we do add more functions later.

**Comment by iquerejeta** [comment]
@KtorZ , do you think the PR is ready to be merged? We are going to start doing some prototyping, so IMO we could merge this as a draft, WDYT?

**Comment by KtorZ** [comment]
@iquerejeta we can add it to the agenda for next session (cc @mangelsjover) for review. 

I am yet to go through the comments and discussions that happened as part of this one, but I can already make a few points that would need to be fixed before we merge it:

- Change the name & folder to CIP-0381
- Rename _Reason for choosing BLS12_381_  as _Rationale_ and move it after the _Specification_ section. 
- Rename _Moving to Proposed_ as _Path to Proposed_ and consider perhaps having a _Path to Active_ as well (though for this one, I don't see much beyond seeing this actually implemented in Plutus and released in an upcoming update). 
- I am not quite sure about the purpose of the _Some operation count_ section? It seems to sits in-between the realm of recommendations for some of considered use-cases and a justification of the use of pairing in those situations; so perhaps this last section ought to be nested within the `Rationale` section?

**Comment by iquerejeta** [comment]
I've included your suggestions. What I'm not entirely convinced is about the `Some operation count`. I've removed that, and added the operation count in the motivation (beginning of the document). Do you think it fits well there? Happy to make further changes, and thanks for looking into it 👍

**Review by KtorZ — APPROVED** [review]

**Review by rphair — APPROVED** [review]
author has completed changes requested in https://github.com/cardano-foundation/CIPs/pull/220#issuecomment-1174808014.

**Comment by dcoutts** [comment]
In addition to security audits on the proposed implementation, it is important that we review the inclusion of the library for practical issues of build systems, and long term maintainability.

In particular in this case the library will be used in the chain format and in chain verification. This means we have special constraints that the format and verification logic must _never change_. Or more specifically: it must be possible _forever_ to be able to verify the existing uses on the chain. So even if there are upgrades or format changes in future, it must still be possible to decode and verify the old uses. This is not a constraint that most software has, and so many libraries are not designed to work within that constraint.

The proposed implementation is https://github.com/supranational/blst

* The implementation is in C and assembly for x86 and ARM v8. This is good for the existing Haskell implementation of the Cardano node because integrating with C libraries is relatively straightforward, it's well supported by the build and CI system and does not pull in too many extra dependencies. (Contrast for example if it were a Rust library where we would have serious practical problems on this front.)
* The implementation appears to have been designed with blockchain implementations in mind. This is a good sign for the long term maintainability because it probably means that the library authors will continue to support the existing format and semantics even if there are changes or improvements.
* The implementation is claimed to be compliant with draft IETF specifications. There is a risk that the specs may change before being declared final, and the library may be updated to follow. There is a risk that the Cardano community will have to support the old version forever. Though given the point above, it's probably the case that library updates would still provide compatibility with the current IETF drafts and serialisation formats.

So overall this seems like something the core development team and Cardano community could support long term without too high a cost. Though we should be aware of the risk that we may have to support an old version of the library, if the library gets changed in incompatible ways.

To ensure no compatibility surprises, it is worth considering forking the repository at a specific commit/version and building the node using that version. This is to guarantee the version remains available. Then for any future updates, the fork repo could be updated to a new version explicitly, with appropriate compatibility checks to ensure the existing on-chain uses are still compatible.

**Comment by iquerejeta** [comment]
> To ensure no compatibility surprises, it is worth considering forking the repository at a specific commit/version and building the node using that version.

I'm not sure I see what is the benefit of forking here. This would only be useful if it happened that the repo maintainers overwrote history, right? As long as that does not happen, we are good sticking to a particular version/commit of the original repo without the need to fork it.

**Pull request merged**
